src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+	recurse_pg_depend(target,
+					  &follow_get_dependencies_for_object,
+					  &apply_add_to_target_list,
+					  &dependencyList);
+	return dependencyList;
+}
+
+
+/*
+ * IsObjectAddressInList is a helper function that can check if an ObjectAddress is
+ * already in a (unsorted) list of ObjectAddresses
+ */
+static bool
+IsObjectAddressInList(const ObjectAddress *findAddress, List *addressList)
+{
+	ListCell *addressCell = NULL;
+	foreach(addressCell, addressList)
+	{
+		ObjectAddress *currentAddress = (ObjectAddress *) lfirst(addressCell);
+
+		/* equality check according as used in postgres for object_address_present */
+		if (findAddress->classId == currentAddress->classId && findAddress->objectId ==
+			currentAddress->objectId)
+		{
+			if (findAddress->objectSubId == currentAddress->objectSubId ||
+				currentAddress->objectSubId == 0)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * SupportedDependencyByCitus returns whether citus has support to distribute the object
+ * addressed.
+ */
+static bool
+SupportedDependencyByCitus(const ObjectAddress *address)
+{
+	/*
+	 * looking at the type of a object to see if we know how to create the object on the
+	 * workers.
+	 */
+	switch (getObjectClass(address))
+	{
+		case OCLASS_SCHEMA:
+		{
+			return true;
+		}
+
+		default:
+		{
+			/* unsupported type */
+			return false;
+		}
+	}
+
+	/*
+	 * all types should have returned above, compilers complaining about a non return path
+	 * indicate a bug in the above switch. Missing returns or breaking code flow should be
+	 * addressed in the switch statement, preferably not here.
+	 */
+}
+
+
+/*
+ * IsObjectAddressOwnedByExtension returns whether or not the object is owned by an
+ * extension. It is assumed that an object having a dependency on an extension is created
+ * by that extension and therefore owned by that extension.
+ */
+static bool
+IsObjectAddressOwnedByExtension(const ObjectAddress *target)
+{
+	Relation depRel = NULL;
+	ScanKeyData key[2] = { 0 };
+	SysScanDesc depScan = NULL;
+	HeapTuple depTup = NULL;
+	bool result = false;
+
+	depRel = heap_open(DependRelationId, AccessShareLock);
+
+	/* scan pg_depend for classid = $1 AND objid = $2 using pg_depend_depender_index */
+	ScanKeyInit(&key[0], Anum_pg_depend_classid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->classId));
+	ScanKeyInit(&key[1], Anum_pg_depend_objid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->objectId));
+	depScan = systable_beginscan(depRel, DependDependerIndexId, true, NULL, 2, key);
+
+	while (HeapTupleIsValid(depTup = systable_getnext(depScan)))
+	{
+		Form_pg_depend pg_depend = (Form_pg_depend) GETSTRUCT(depTup);
+		if (pg_depend->deptype == DEPENDENCY_EXTENSION)
+		{
+			result = true;
+			break;
+		}
+	}
+
+	systable_endscan(depScan);
+	relation_close(depRel, AccessShareLock);
+
+	return result;
+}
+
+
+/*
+ * OrderObjectAddressListInDependencyOrder given a list of ObjectAddresses return a new
```

Comment:
Maybe ` list of the same ObjectAddresses ordered on dependency order with the objects without dependencies first.` > ` list of the same ObjectAddresses ordered on dependency order where dependencies precedes the corresponding object in the list `